Client/Frontend/Library/BookmarkDetailPanel.swift

```diff
@@ -0,0 +1,338 @@
+                    return result.isFailure ? deferMaybe(BookmarkDetailPanelError()) : succeed()
+                })
+            }
+        } else {
+            guard let bookmarkNodeGUID = self.bookmarkNodeGUID else {
+                return deferMaybe(BookmarkDetailPanelError())
+            }
+
+            if bookmarkNodeType == .bookmark {
+                return profile.places.updateBookmarkNode(guid: bookmarkNodeGUID, parentGUID: parentBookmarkFolder.guid, title: bookmarkItemOrFolderTitle, url: bookmarkItemURL)
+            } else if bookmarkNodeType == .folder {
+                return profile.places.updateBookmarkNode(guid: bookmarkNodeGUID, parentGUID: parentBookmarkFolder.guid, title: bookmarkItemOrFolderTitle)
+            }
+        }
+
+        return deferMaybe(BookmarkDetailPanelError())
+    }
+
+    // MARK: UITableViewDataSource | UITableViewDelegate
+
+    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
+        tableView.deselectRow(at: indexPath, animated: false)
+
+        guard indexPath.section == BookmarkDetailSection.folder.rawValue else {
+            return
+        }
+
+        if isFolderListExpanded, let item = bookmarkFolders[safe: indexPath.row] {
+            parentBookmarkFolder = item.folder
+        }
+
+        isFolderListExpanded = !isFolderListExpanded
+
+        tableView.reloadSections(IndexSet(integer: indexPath.section), with: .automatic)
+    }
+
+    func numberOfSections(in tableView: UITableView) -> Int {
+        return BookmarkDetailSection.count
+    }
+
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        if section == BookmarkDetailSection.fields.rawValue {
+            switch bookmarkNodeType {
+            case .bookmark:
+                return 2
+            case .folder:
+                return 1
+            default:
+                return 0
+            }
+        } else if section == BookmarkDetailSection.folder.rawValue {
+            if isFolderListExpanded {
+                return bookmarkFolders.count
+            } else {
+                return 1
+            }
+        }
+
+        return 0 // Should not happen.
+    }
+
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        // Handle folder selection cells.
+        guard indexPath.section == BookmarkDetailSection.fields.rawValue else {
+            guard let cell = tableView.dequeueReusableCell(withIdentifier: BookmarkDetailFolderCellIdentifier, for: indexPath) as? OneLineTableViewCell else {
+                return super.tableView(tableView, cellForRowAt: indexPath)
+            }
+
+            cell.imageView?.image = UIImage(named: "bookmarkFolder")?.createScaled(CGSize(width: BookmarkDetailPanelUX.FolderIconSize, height: BookmarkDetailPanelUX.FolderIconSize))
```

Comment:
Maybe change definition of `BookmarkDetailPanelUX.FolderIconSize = CGSize(width: xx, height:xx)`, so then createScaled(BookmarkDetailPanelUX.FolderIconSize)

Reply:
Done.